### PR TITLE
ZDC unpacker into castor, backport to 80X

### PR DESCRIPTION
**Pull request by cferraio**
Hi all,

I've moved ZDC unpacking from hcal to digi. The same change has previously been made for 75X and 81X. The delay for 80X was due to ongoing discussion about hardware updates that have been put on hold until after the pPb run.

I've also added digi->reco functionality for data unpacked through the new ZDC unpacker.

-Chris

P.S. For those following ZDC software discussions - this PR is only related to the "old" ZDC subdets, and does not address a future PR related to the new RPD subdet.

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cferraio 
is the code in this PR identical to what we have in 81X?
If some features are missing from 81X, they have to appear there first.

**Comment by cferraio** [comment]
@slava77 yes it should be. i also included digi->reco in the 81X PR, and the unpacker code is the same.

**Comment by mandrenguyen** [comment]
@slava77 @cvuosalo :  What's the timescale for signing off on this PR?

File: RecoLocalCalo/HcalRecProducers/src/ZdcSimpleReconstructor.h [hunk 1]
```diff
@@ -31,7 +32,8 @@
       DetId::Detector det_;
       int subdet_;
       HcalOtherSubdetector subdetOther_;
-      edm::EDGetTokenT<ZDCDigiCollection> tok_input_;
+      edm::EDGetTokenT<ZDCDigiCollection> tok_input_hcal;
+      edm::EDGetTokenT<ZDCDigiCollection> tok_input_castor;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/src/ZdcSimpleReconstructor.h]: there is somewhat trivial difference between this and 81X: the two lines with tok_input_hcal and tok_input_castor are flipped.
Please keep a consistent backport.

**Comment by slava77** [comment]
backport of #14082

File: EventFilter/CastorRawToDigi/interface/ZdcUnpacker.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+private:
+  int sourceIdOffset_; ///< number to subtract from the source id to get the dcc id
+  int startSample_; ///< first sample from fed raw data to copy 
+  int endSample_; ///< last sample from fed raw data to copy (if present)
+  int expectedOrbitMessageTime_; ///< Expected orbit bunch time (needed to evaluate time differences)
+  std::set<HcalElectronicsId> unknownIds_,unknownIdsTrig_; ///< Recorded to limit number of times a log message is generated
+  int mode_;
+};
+
+#endif // ZdcUnpacker_h_included
```
Comment (slava77) [inline EventFilter/CastorRawToDigi/interface/ZdcUnpacker.h]: add a newline (to be consistent with 81X)

File: RecoLocalCalo/HcalRecProducers/src/ZdcSimpleReconstructor.cc [hunk 3]
```diff
@@ -20,7 +20,9 @@ ZdcSimpleReconstructor::ZdcSimpleReconstructor(edm::ParameterSet const& conf):
   det_(DetId::Hcal),
   dropZSmarkedPassed_(conf.getParameter<bool>("dropZSmarkedPassed"))
 {
-  tok_input_ = consumes<ZDCDigiCollection>(conf.getParameter<edm::InputTag>("digiLabel"));
+  tok_input_castor = consumes<ZDCDigiCollection>(conf.getParameter<edm::InputTag>("digiLabelcastor"));
+  tok_input_hcal = consumes<ZDCDigiCollection>(conf.getParameter<edm::InputTag>("digiLabelhcal"));
```
Comment (cferraio) [inline RecoLocalCalo/HcalRecProducers/src/ZdcSimpleReconstructor.cc]: hi @slava77, I'm about to make a PR to 81X to fix this bug. Should I include edits to the other two files (EventFilter/CastorRawToDigi/interface/ZdcUnpacker.h and RecoLocalCalo/HcalRecProducers/src/ZdcSimpleReconstructor.h) as well, to make them match this 80X PR? That way only one PR for 81X and no more changes to this one.

**Comment by cferraio** [comment]
Added a line to ZdcUnpacker.h. Fixed other two issues (minor bug and line flip) in PR for 81X (PR #16196).

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16128 4ce669c
- code changes are as described
- jenkins tests pass and comparisons show no differences
- local test of 2015 HI data (run 262694 LS 100 ZDC was read out in Castor FEDs) shows that ZDC data is now reconstructed appropriately 
wfhihardprobes2015_centrality_zdc

@davidlange6 
Note that new data product is coming from this PR in pp workflows as well (standard raw2digi). So, integration in 80X production should be coordinated with a processing version or acquisition era change.

**Comment by cferraio** [comment]
hi @slava77 and @davidlange6, is there anything else that needs to be done? anything I could look into? thanks.

**Comment by davidlange6** [comment]
+1

**Pull request merged**